### PR TITLE
Initial steps to back context managers by the `contextManager` interface

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR follows up on the work in #22857 to make context managers be backed by a `contextManager` interface. For the time being, we do not want to _require_ the context manager interface, only _warn_ about not implementing it. This is done using the `implements interface` primitive that was added for #22857, and by adding a call to `chpl__verifyTypeContext` with a compiler warning to each context manager expression. 

There is some mismatch between context managers and the design of interfaces as they are today.

1. The first problem is that different context manager `enterContext` methods return different types, depending on the resource being managed. Thus, the interface must have the ability to support all return types gracefully. This is solved in a relatively "canonical" way by using an associated type: each context manager provides a `contextReturnType` parenless type procedure that is used to determine the expected type for `enterContext`. However, the goal is for a warning-only migration, and we don't want to introduce an additional burden on the users; thus, the compiler will automatically infer `contextReturnType` for `contextManager` where it can.
2. The second problem is that the context managers can also return by different intents. Unlike the return type issue, this cannot be solved by an existing interface feature. I therefore introduce a pragma, `ifc any return intent` to allow any return intent to satisfy the interface's witness. This pragma interacts poorly with constraint-generic functions (because those are resolved once, and thus cannot be differentiated based on the return type of the actual witness), and thus `contextManager` is banned as a constraint in CG functions. 

While there, it fixes a bug in which nested aggregate types caused segmentation faults when an `: interface` syntax was used. This was because the `implements` statement for the child class was inserted into the body of the parent class, which only expected definitions (not a statement). Since flattening classes happens pretty early on, and moves child classes into the same scope as the parent class, the solution seemed to simply insert implements statements in module scope, like `FlattenClasses` does. 

It's unclear how the interface will handle the `contextManager` range of features in the long-term, since CM allows return intent overloading and (as described above) differentiated return intents. However, since this isn't strictly necessary for 2.0, maybe that's fine?

Reviewed by @vasslitvinov -- thanks!

## Testing
- [x] paratest

### CODE REVIEW

File: compiler/AST/AggregateType.cpp [hunk 1]
```diff
@@ -3113,7 +3113,7 @@ void AggregateType::addClassToHierarchy(std::set<AggregateType*>& localSeen) {
 
       auto ifcActuals = new CallExpr(PRIM_ACTUALS_LIST, new SymExpr(implementFor));
       auto istmt = ImplementsStmt::build(isym->name, ifcActuals, nullptr);
-      this->symbol->defPoint->insertAfter(istmt);
+      this->symbol->getModule()->block->insertAtTail(istmt);
```
Comment (vasslitvinov) [inline compiler/AST/AggregateType.cpp:3116]: I would like to see an explanation of why this change is necessary.

File: compiler/AST/interfaces.cpp [hunk 2]
```diff
@@ -71,6 +72,8 @@ DefExpr* InterfaceSymbol::buildDef(const char* name,
 
   if (gHashable == nullptr && name == astr("hashable")) {
     gHashable = isym;
+  } else if (gContextManager == nullptr && name == astr("contextManager")) {
+    gContextManager = isym;
   }
```
Comment (vasslitvinov) [inline compiler/AST/interfaces.cpp:77]: Do the checks `gHashable == nullptr` and `gContextManager == nullptr` ensure that gHashable and gContextManager will refer to the predefined interfaces?

I think `!strcmp(name, "hashable")` is faster (and more direct) than `name == astr("hashable")`.

File: compiler/resolution/interfaceResolution.cpp [hunk 3]
```diff
@@ -833,6 +844,77 @@ static BlockStmt* createHolderBlock(FnSymbol* wrapFn, ImplementsStmt* istm) {
   return holder;
 }
 
+static bool foundTargetFunctionValid(FnSymbol* tg, bool generatedOnly) {
+  (void) generatedOnly;
```
Comment (vasslitvinov) [inline compiler/resolution/interfaceResolution.cpp:848]: I'd like to see a comment why this line `(void) generatedOnly;` is needed. Ideally, it is short enough to fit on the same line.

In the function name, I prefer `IsValid` instead of simply `Valid`. Better yet, simplify to `targetFunctionIsValid`. Assuming that makes sense. Otherwise I would like to see a comment explaining what the function is checking.

File: doc/rst/language/spec/statements.rst [hunk 4]
```diff
@@ -978,17 +978,17 @@ context managers. The syntax of the manage statement is given by
     expression
 
 Classes or records that wish to be used as context managers must
-define two special methods. The code sample below turns a record
-type named ``IntWrapper`` into a context manager and then uses it
-in a manage statement.
+define two special methods, and implement the ``contextManager`` interface.
+The code sample below turns a record type named ``IntWrapper`` into a context
+manager and then uses it in a manage statement.
```
Comment (vasslitvinov) [inline doc/rst/language/spec/statements.rst:983]: I suggest that "implement CM" is fundamental and "define two methods" is a consequence. Also I suggest avoiding the term "special" as I do not know what that means, formally speaking. Also "turn IntWrapper into a context manager" does not make sense because IntWrapper is intended as a CM from the beginning. So:

[the first line unchanged] implement the CM interface by defining two methods enterContext and exitContext. The example below defines a context manager IntWrapper and uses it in a `manage` statement.

File: test/constrained-generics/contextManager/PREDIFF [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+command='\|CHPL_HOME/modules|s/:[0-9:]*:/:nnnn:/'
```
Comment (vasslitvinov) [inline test/constrained-generics/contextManager/PREDIFF:3]: I suggest removing `:` from inside the brackets `[0-9:]`

File: test/constrained-generics/contextManager/infer-return-type.chpl [hunk 6]
```diff
@@ -0,0 +1,16 @@
+record R : contextManager {
+  proc enterContext() ref {
+    writeln("entering");
+    return none;
```
Comment (vasslitvinov) [inline test/constrained-generics/contextManager/infer-return-type.chpl:4]: How is this supposed to work, returning a literal by `ref` ?  To me this should be a compilation error.

File: compiler/AST/AggregateType.cpp (hunk 1 above)
Comment (DanilaFe) [reply compiler/AST/AggregateType.cpp:3116]: Updated the OP to include this:

>  While there, it fixes a bug in which nested aggregate types caused segmentation faults when an `: interface` syntax was used. This was because the `implements` statement for the child class was inserted into the body of the parent class, which only expected definitions (not a statement). Since flattening classes happens pretty early on, and moves child classes into the same scope as the parent class, the solution seemed to simply insert implements statements in module scope, like `FlattenClasses` does.

File: compiler/AST/interfaces.cpp (hunk 2 above)
Comment (DanilaFe) [reply compiler/AST/interfaces.cpp:77]: > Do the checks gHashable == nullptr and gContextManager == nullptr ensure that gHashable and gContextManager will refer to the predefined interfaces?

Yup!

File: test/constrained-generics/contextManager/infer-return-type.chpl (hunk 6 above)
Comment (DanilaFe) [reply test/constrained-generics/contextManager/infer-return-type.chpl:4]: It's not an error because of: https://github.com/chapel-lang/chapel/issues/23159

I'll see if I can generate the "wrappers" for void-typed functions, though.

File: doc/rst/language/spec/statements.rst (hunk 4 above)
Comment (DanilaFe) [reply doc/rst/language/spec/statements.rst:983]: I've taken a stab at rewording this somewhat.

File: test/constrained-generics/contextManager/infer-return-type.chpl (hunk 6 above)
Comment (DanilaFe) [reply test/constrained-generics/contextManager/infer-return-type.chpl:4]: Turns out, return-type inference + any-return-intent means I don't need `none` at all, and `void` works natively :)

**Review by vasslitvinov — APPROVED** [review]
